jlm/llvm/opt/alias-analyses/PointerObjectSet.cpp

```diff
@@ -958,6 +972,273 @@ PointerObjectConstraintSet::DrawSubsetGraph(util::GraphWriter & writer) const
+      {
+        predecessorEquivalenceLabels[successorSCC] = SEVERAL_PREDECESSORS;
+      }
+    }
+  }
+
+  return equivalenceSetLabels;
+}
+
+size_t
+PointerObjectConstraintSet::DoOfflineVariableSubstitution()
+{
+  // Performing unification on direct nodes relies on all subset edges being known offline.
+  // This is only safe if no more constraints are added to the node in the future.
+  ConstraintSetFrozen_ = true;
+
+  // For each PointerObject v, creates two nodes: n(v) and n(*v), and creates edges between them
+  auto subsetGraph = CreateOVSSubsetGraph();
+  auto totalNodeCount = std::get<0>(subsetGraph);
+  auto & successors = std::get<1>(subsetGraph);
+  auto & isDirectNode = std::get<2>(subsetGraph);
+
+  auto GetSuccessors = [&](size_t node)
+  {
+    return successors[node].Items();
+  };
+
+  // Output vectors from Tarjan's SCC algorithm
+  std::vector<size_t> sccIndex;
+  std::vector<size_t> topologicalOrder;
+  auto numSccs = util::FindStronglyConnectedComponents(
+      totalNodeCount,
+      GetSuccessors,
+      sccIndex,
+      topologicalOrder);
+
+  // Find out which SCCs only contain direct nodes
+  std::vector<bool> sccHasDirectNodesOnly(numSccs, true);
+  for (size_t node = 0; node < totalNodeCount; node++)
+  {
+    if (!isDirectNode[node])
+      sccHasDirectNodesOnly[sccIndex[node]] = false;
+  }
+
+  // Give each SCC an equivalence set label
+  auto equivalenceSetLabels = AssignOVSEquivalenceSetLabels(
+      successors,
+      numSccs,
+      sccIndex,
+      topologicalOrder,
+      sccHasDirectNodesOnly);
+
+  // Finally unify all PointerObjects with equal equivalence label
+  size_t numUnifications = 0;
+  std::vector<std::optional<PointerObjectIndex>> unificationRoot(
+      equivalenceSetLabels.size(),
+      std::nullopt);
+
+  for (PointerObjectIndex i = 0; i < Set_.NumPointerObjects(); i++)
+  {
+    const auto equivalenceSetLabel = equivalenceSetLabels[sccIndex[i]];
+
+    // If other nodes with the same equivalence set label have been seen, unify it with i
+    if (unificationRoot[equivalenceSetLabel])
+    {
+      unificationRoot[equivalenceSetLabel] =
+          Set_.UnifyPointerObjects(i, *unificationRoot[equivalenceSetLabel]);
+      numUnifications++;
+    }
+    else
+      unificationRoot[equivalenceSetLabel] = i;
+  }
+
+  return numUnifications;
+}
+
+size_t
+PointerObjectConstraintSet::NormalizeConstraints()
+{
+  // The new list of constraint, preserving the order of constraints that are not deleted.
+  std::vector<ConstraintVariant> newConstraints;
+
+  // Sets used to avoid adding duplicates of any constraints
+  std::set<std::pair<PointerObjectIndex, PointerObjectIndex>> addedSupersetConstraints;
+  std::set<std::pair<PointerObjectIndex, PointerObjectIndex>> addedStoreConstraints;
+  std::set<std::pair<PointerObjectIndex, PointerObjectIndex>> addedLoadConstraints;
+  std::set<std::pair<PointerObjectIndex, const CallNode *>> addedCallConstraints;
```

Comment:
Why do you use `std::set` instead of `std::unordered_set` or `HashSet`?

Reply:
Good point. Originally I tried to make all the Constraint variants, e.g. `StoreConstraint`, implement `operator <`, which would allow me to put them directly into a set. Implementing the comparison operator feels nicer than overloading `std::hash`. I ended up not liking it that much and went over to using `std::pair`, which does overload `std::hash`, so I will switch to `unordered_set`, i.e. `util::HashSet`.

Reply:
Actually, I did try using `HashSet`. It turns out `std::pair` doesn't have an overload for `std::hash`, which makes a lot of people very angry and is widely regarded as a bad move. I will create a separate PR with a workaround, if we think it is worth it.